src/main/scala/fi/oph/koski/kyselyt/KyselyTulosRepository.scala

```diff
@@ -0,0 +1,96 @@
+package fi.oph.koski.kyselyt
+
+import com.typesafe.config.Config
+import fi.oph.koski.config.Environment
+import fi.oph.koski.log.Logging
+import software.amazon.awssdk.auth.credentials.{AwsSessionCredentials, StaticCredentialsProvider}
+import software.amazon.awssdk.core.sync.RequestBody
+import software.amazon.awssdk.http.ContentStreamProvider
+import software.amazon.awssdk.regions.Region
+import software.amazon.awssdk.services.s3.S3Client
+import software.amazon.awssdk.services.s3.model.{CreateBucketRequest, GetObjectRequest, GetUrlRequest, PutObjectRequest}
+import software.amazon.awssdk.services.s3.presigner.S3Presigner
+import software.amazon.awssdk.services.s3.presigner.model.GetObjectPresignRequest
+
+import java.io.InputStream
+import java.net.URI
+import java.time.Duration
+import java.util.UUID
+import scala.jdk.CollectionConverters._
+
+class KyselyTulosRepository(config: Config) extends Logging {
+  val useAWS = Environment.isServerEnvironment(config)
+  lazy val region: Region = Region.of(config.getString("kyselyt.s3.region"))
+  lazy val bucketName: String = config.getString("kyselyt.s3.bucket")
+  lazy val presignDuration: Duration = config.getDuration("kyselyt.s3.presignDuration")
+  lazy val endpointOverride: URI = URI.create(config.getString("kyselyt.s3.endpoint"))
+
+  val s3: S3Client = {
+    val awsS3 = S3Client.builder().region(region)
+    val awsOrLocalS3 = if (useAWS) awsS3 else {
+      logger.warn("Using Localstack for S3")
+      awsS3
+        .endpointOverride(endpointOverride)
+        .credentialsProvider(localstackCredentialsProvider)
+    }
+    awsOrLocalS3.build()
+  }
+
+  if (!useAWS) {
+    // Bucketin automaattinen luonti ainoastaan Localstackin kanssa
+    createBucketIfDoesNotExist
+  }
+
+  def put(queryId: UUID, name: String, provider: ContentStreamProvider, contentType: String): Unit = {
+    val key = objectKey(queryId, name)
+    val request = PutObjectRequest.builder()
+      .bucket(bucketName)
+      .key(key)
+      .contentType(contentType)
+      .metadata(mapAsJavaMap(Map {
+        "query" -> queryId.toString
+      }))
+      .build()
+    val requestBody = RequestBody.fromContentProvider(provider, contentType)
+    logger.info(s"Put result to S3: ${s3.utilities().getUrl(GetUrlRequest.builder().bucket(bucketName).key(key).build())} ($contentType)")
+    s3.putObject(request, requestBody)
+  }
+
+  def getPresignedDownloadUrl(queryId: UUID, name: String): String = {
+    val key = objectKey(queryId, name)
+    val awsPresigner = S3Presigner.builder()
+    val presigner = if (useAWS) awsPresigner.build() else awsPresigner
```

Comment:
vähän hämäävä if else -rivitys, joutuu päässään parsimaan, ajetaanko 3 erillistä riviä myös awsPresigner.build():in lopputulokselle vai ainoastaan awsPresigner:lle.